bpf/bpf_host.c

```diff
@@ -916,6 +916,11 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 {
 	__u16 proto;
 
+	/* Pass vlan tagged packets to the stack - kernel will strip vlan tag and we'll receive it back on proper interface */
+	if (ctx->vlan_present) {
```

Comment:
It's in the fast path, so let's do the check only if `from_host == false`.

Reply:
You're right. And even more. I think that we need to skip packet processing on egress in to_netdev function. PR update coming up.

Reply:
> skip packet processing on egress in to_netdev function

What's the packet flow for egress in terms of VLAN devices in the kernel?

Reply:
wan0 - physical interface, wan0.4000 - vlan interface attached to wan0
On ingress packet will appear with vlan tag on wan0 first and then on wan0.4000 without tag.
On egress packet will appear on wan0.4000 first without vlan tag and then on wan0 with tag.

Reply:
In:
wan0:
```
52:54:00:63:7b:cc > 52:54:00:12:34:56, ethertype 802.1Q (0x8100), length 72: vlan 4000, p 0, ethertype IPv4 (0x0800)
10.118.13.1.35330 > 10.118.11.20.5678
```
wan0.4000:
```
52:54:00:63:7b:cc > 52:54:00:12:34:56, ethertype IPv4 (0x0800), length 68
10.118.13.1.35330 > 10.118.11.20.5678
```

Out:
wan.4000:
```
52:54:00:12:34:56 > 52:54:00:12:34:58, ethertype IPv4 (0x0800)
10.118.13.1.35330 > 10.244.2.67.5678
```
wan0:
```
52:54:00:12:34:56 > 52:54:00:12:34:58, ethertype 802.1Q (0x8100), length 72: vlan 4000, p 0, ethertype IPv4 (0x0800)
10.118.13.1.35330 > 10.244.2.67.5678
```

Reply:
OK, then it makes sense to skip the processing.